feat: 쿠폰 엔티티 생성

PNoahKR opened the pull request:
## 📝작업 내용

- Coupons 엔티티 생성(Coupon으로 생성시 오류발생)
- UserCoupon 엔티티 생성
- BaseEntity 생성 : created_at, updated_at 등 공용 엔티티

### 스크린샷

## 💬리뷰 요구사항

## 🔗참고 링크(선택)

## ⏰기한

- 2024-09-12

PNoahKR committed:
feat: 쿠폰 기능 엔티티 생성
- Coupons 엔티티 생성(Coupon으로 생성시 오류발생)
- UserCoupon 엔티티 생성
- BaseEntity 생성 : created_at, updated_at 등 공용 엔티티

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/UserCoupon.java:
```diff
@@ -0,0 +1,47 @@
+    @JoinColumn(name = "coupon_id")
+    private Coupons couponId;
+
+    @Column(name = "is_used", nullable = false)
+    private boolean isUsed = false; //쿠폰 사용 여부
+
```
isUse = false 코드는 어떤것 때문에 초기화를 해주는건가요?
객체 생성때라면 차라리 생성자코드 안에 넣는건 어떠신가여

PNoahKR replied:
@evan-song2 생성자에 넣도록 하겠습니다!

PNoahKR replied:
@evan-song2 [9096f27](https://github.com/team-fashionmall/fashionmall/pull/6/commits/9096f2771ae7948995c36a8deb7f163c238f1cb6) 생성자에 추가

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/UserCoupon.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void UseCoupon() {
+        if (isUsed) {
+            throw new CustomException(ErrorResponseCode.ORDER_NOT_FOUND_COUPON);
+        }
```
예외 처리와 상태 변경의 책임을 분리하는건 어떠실까요? 
유연하고 유지보수 가능하게 만들기 위해 서비스 계층에서 처리하는 방식도 고려해보면 좋을 것 같습니다

evan-song2 replied:
메서드명 수정해주세요

PNoahKR replied:
@evan-song2 상태변경 메소드는 markAsUsed로 변경해 유지하고 예외 처리는 서비스 계층에서 구현하도록 하겠습니다!
```java
public void markAsUsed() {
        isUsed = true;
        usedAt = LocalDateTime.now();
    }
``` 
+ [9096f27](https://github.com/team-fashionmall/fashionmall/pull/6/commits/9096f2771ae7948995c36a8deb7f163c238f1cb6)

evan-song2 replied:
@PNoahKR 
아 제가 메서드명 변경해달라고 했던부분은 메서드명이 대문자로 시작해서였습니다
매서드명은 기존안이 더 좋아보입니다

PNoahKR replied:
@evan-song2 [ce88512](https://github.com/team-fashionmall/fashionmall/pull/6/commits/ce88512c79425b6ca41d52f87c7d7f8154f606a7) 아 대문자 때문이었네요! 다시 기존안으로 변경했습니다!

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/Coupons.java:
```diff
@@ -0,0 +1,86 @@
+
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(name = "coupons")
+public class Coupons extends BaseEntity {
```
Coupon이 생성 안된 이유는 어떤것 때문인가요?

PNoahKR replied:
@evan-song2 전날 테스트 및 여러가지 설정을 만질때 안되던 부분이 지금 Coupon으로도 실행이 가능해져 Coupon으로 통일시키도록 하겠습니다!

PNoahKR replied:
@evan-song2 [9096f27](https://github.com/team-fashionmall/fashionmall/pull/6/commits/9096f2771ae7948995c36a8deb7f163c238f1cb6) 클래스명 변경

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/Coupons.java:
```diff
@@ -0,0 +1,86 @@
+    @Column(name = "disouct_type", nullable = false)
+    private DiscountType discountType; //쿠폰 할인 타입(RATE: 정률, AMOUNT: 정액)
+
+    @Column(name = "discount_value", nullable = false)
+    private int discountValue; //할인 값
+
+    @Column(name = "min_purchase_price", nullable = false)
+    private int minPurchasePrice; //최소 주문 금액
+
+    @Column(name = "max_discount_pirce")
+    private Integer maxDiscountPrice; //최대 할인 금액(정률 할인),
```
int와 Integer를 같이 사용하신 이유는 DB default값 존재에 때라 달라지는건가요?

PNoahKR replied:
@evan-song2 null이 허용되는 곳에만 Integer를 넣었는데 모두 Integer로 통일하는 편이 좋을까요?

evan-song2 replied:
@PNoahKR 일반적으로 엔티티는 원시타입보단 객체형태로 선언하긴 하는데
not null조건이 있다면 꼭 관례를 따르진 않아도 된다고 생각해서 나쁘지 않아보이네요~

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/Coupons.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    private void validateMaxDiscountPrice() {
+        if (this.discountType == DiscountType.RATE && (maxDiscountPrice == null || maxDiscountPrice <= 0)) {
+            throw new CustomException(ErrorResponseCode.BAD_REQUEST);
+        }
+        if (this.discountType == DiscountType.AMOUNT) {
+            this.maxDiscountPrice = null;
+        }
```
1. 최대할인금액은 정률일때만 저장한다는 정책이였나요?
2. 엔티티 생성자 내에서 검증을 하는건 유연성이 떨어진다고 생각합니다. 
3. maxDiscountPrice 세팅해주는건 엔티티 생성할때 해줘도 좋을것 같습니다

PNoahKR replied:
@evan-song2 
1. 정률 할인 시 최대할인금액(쿠폰으로 할인되는 값의 한계값)을 정하기위해 정률시에만 저장하는 정책이었습니다.
2. 검증도 생성시 서비스 단에서 검증하도록 하겠습니다.
검증 메서드는 엔티티에 유지하고 생성시 서비스에서 검증 및 세팅하도록 하면 될까요?

PNoahKR replied:
@evan-song2 [9096f27](https://github.com/team-fashionmall/fashionmall/pull/6/commits/9096f2771ae7948995c36a8deb7f163c238f1cb6) 생성자에 검증 메소드 제거

evan-song2 replied:
@PNoahKR 
상황에 따라 엔티티 생성 조건이 달라질 수 있는데
생성자내부에 넣어버리면 엔티티 생성에 제약이 생기는것 같아보입니다.

생성하기 전에 확인을 하고 생성자를 호출하는게 더 유연해보인다고 생각합니다

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/CouponApplication.java:
```diff
@@ -7,9 +7,7 @@
 @SpringBootApplication
 @ComponentScan(basePackages = {"com.fashionmall.coupon", "com.fashionmall.common"})
 public class CouponApplication {
```
Auditing 설정이 빠져있는것 같아 보이는데 BaseEntity 정상 동작 하나요??

PNoahKR replied:
@evan-song2 
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/39370361-6afe-45b8-af69-2b8fc10cf7b0">
created_at, updated_at이 정상적으로 입력돼 나오는데 정상 동작하는것으로 보입니다..!

evan-song2 replied:
@PNoahKR 확인 감사합니다~
PrePersist 는 자바에서 제공해주는 어노테이션이라 어디팅설정 없이도 사용 가능한가보네요 👍🏻

PNoahKR committed:
refactor: 엔티티명 변경 및 코드 수정
(Coupon)
- Coupons -> Coupon
- 생성자에서 MaxDiscountPrice 검증 메소드 제거
(UserCoupon)
- UseCoupon -> markAsUsed 메소드명 변경 및 예외 처리와 상태 변경 분리
- isUsed 생성자에서 초기값 설정

evan-song2 commented on module-coupon-service/src/main/java/com/fashionmall/coupon/entity/UserCoupon.java:
```diff
@@ -24,23 +22,21 @@ public class UserCoupon extends BaseEntity {
     @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "coupon_id")
-    private Coupons couponId;
+    private Coupon couponId;
```
Coupon이라는 객체를 의존하고 있으니 필드명은 couponId보단 coupon이 나아보입니다

PNoahKR replied:
@evan-song2 [7335926](https://github.com/team-fashionmall/fashionmall/pull/6/commits/7335926e90c354bebb23137bb332d71f99e8f799) 수정 완료했습니다!

PNoahKR committed:
refactor: 쿠폰 사용 메서드명
- useCoupon으로 변경

PNoahKR committed:
refactor: 필드명 변경
- couponID -> coupon으로 변경

evan-song2 reviewed (approved):
고생하셨습니다~
